test/fail_compilation/fix21868.d

```diff
@@ -0,0 +1,14 @@
+/* REQUIRED_ARGS: -preview=dip1000
+TEST_OUTPUT:
+---
+fail_compilation/fix21868.d(12): Error: scope variable `p` may not be returned
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=21868
+
+@safe ref int test(ref scope return int* p)
```

Comment:
Since the function is returning by `ref`, `p` should be treated as `return ref` and `scope`. From my perspective, it means that this code should behave like:

```d
@safe ref int test(ref return int* p)
{
    return *p;
}
```

but also make sure that `p` doesn't escape by other means than `return ref`.  Why would you even bother to add `return ref` is `scope overrides it? I think that this code should compile.

Reply:
There are two things at play here:

1. the value of p
2. the address of p

Which one is being returned? Those two are independent properties.

Reply:
> Which one is being returned? Those two are independent properties.

Does it matter though? The lifetime of the value is always larger than or equal to the lifetime of the address and, the returned lifetime is currently always that of the address ([it seems](https://forum.dlang.org/post/hncismmrrtpwtavmtvaa@forum.dlang.org)), so ref-return should cover both cases (unless I'm missing something).

Reply:
Conflating two independent properties is never going to work out very well.

Reply:
It's what got us into this particular issue in the first place. Not going to add to it.

Reply:
> Conflating two independent properties is never going to work out very well.

They are not independent. Here's an exercise: try to come up with an example where the test case from this PR actually causes memory corruption in `@safe` code prior to this bug fix. Here's a starting attempt:

```D
@safe:
ref int test(ref scope return int* p) {
    return *p;
}

ref int escape() {
    int x;
    int* ptr = &x;
    return test(ptr); // Error: returning `test(ptr)` escapes a reference to local variable `ptr`
}

void main() {
    escape() = 3; // try to assign to a dangling reference
}
```

To make it pass, I'd need to make the lifetime of `&ptr` larger than the lifetime of `&x`, which you can't do.

Reply:
You can have a ref to a pointer variable, and assign the pointer any value you want. The two are independent, since neither scope nor ref are transitive.